src/appkit.cs

```diff
@@ -27427,4 +27575,265 @@ interface NSStepperTouchBarItem
 		[Export ("customizationLabel")]
 		string CustomizationLabel { get; set; }
 	}
+
+	[Protocol]
+	interface NSTextInputTraits
+	{
+		[Export ("autocorrectionType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType AutocorrectionType { get; set; }
+
+		[Export ("spellCheckingType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType SpellCheckingType { get; set; }
+
+		[Export ("grammarCheckingType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType GrammarCheckingType { get; set; }
+
+		[Export ("smartQuotesType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType SmartQuotesType { get; set; }
+
+		[Export ("smartDashesType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType SmartDashesType { get; set; }
+
+		[Export ("smartInsertDeleteType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType SmartInsertDeleteType { get; set; }
+
+		[Export ("textReplacementType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType TextReplacementType { get; set; }
+
+		[Export ("dataDetectionType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType DataDetectionType { get; set; }
+
+		[Export ("linkDetectionType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType LinkDetectionType { get; set; }
+
+		[Export ("textCompletionType", ArgumentSemantic.Assign)]
+		NSTextInputTraitType TextCompletionType { get; set; }
+	}
+
+	interface INSTextCheckingClient { }
+
+	[Protocol]
+	interface NSTextCheckingClient : NSTextInputTraits, NSTextInputClient
+	{
+		[Abstract]
+		[Export ("annotatedSubstringForProposedRange:actualRange:")]
+		[return: NullAllowed]
+		NSAttributedString AnnotatedSubstring (NSRange range, [NullAllowed] NSRange actualRange);
```

Comment:
In ObjC this is a required method of a protocol that returns a pointer to `NSAttributedString`. Does it make sense to make this `[Internal]` and do a manual implementation mapping a `IntPtr` returned to `NSAttributedString`?

Reply:
Not sure if I get your concern correctly but it's a protocol, so an interface once generated
Making a member internal would not be helpful, even less if required since no one would be able to conform

Reply:
Method name lacks a verb, e.g. `Get...`

Reply:
@spouliot here's the code Pramit is trying to bind:

```
unsafe NSAttributedString AttributedSubstringForProposedRange (NSRange range, [NullAllowed] NSRangePointer* actualRange);
```

Similar to https://github.com/xamarin/xamarin-macios/pull/7295/files#r338313032 he's hiding a pointer here.

A similar API:
```
unsafe CGRect FirstRectForCharacterRange (NSRange range, [NullAllowed] NSRangePointer* actualRange);
```
was bound like this: https://github.com/xamarin/xamarin-macios/blob/master/src/appkit.cs#L19304

So you can use ref/out in those cases because you're dealing with a pointer.